jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/util/NonBindingMethodHolder.java

```diff
@@ -0,0 +1,137 @@
+{
+    private final MethodHandle _methodHandle;
+    private final Object[] _parameters;
+    private final Boolean[] _boundParams;
+    private final int _numParams;
+    private Class<Object> _returnType;
+    private MethodHandle _returnFilter;
+
+    public NonBindingMethodHolder(MethodHandle methodHandle)
+    {
+        _methodHandle = Objects.requireNonNull(methodHandle);
+        _numParams = methodHandle.type().parameterCount();
+        _parameters = new Object[_numParams];
+        _boundParams = new Boolean[_numParams];
+    }
+
+    private int getInternalIndex(int idx)
+    {
+        int index = 0;
+        for (int i = 0; i < _numParams; i++)
+        {
+            if (_boundParams[i] != Boolean.TRUE)
+            {
+                if (index == idx)
+                    return i;
+                index++;
+            }
+        }
+
+        throw new IndexOutOfBoundsException(idx);
+    }
+
+    @Override
+    public Object invoke(Object... args) throws Throwable
+    {
+        insertArguments(false, 0, args);
+        Object o = _methodHandle.invokeWithArguments(_parameters);
+        if (_returnFilter != null)
+            o = _returnFilter.invoke(o);
+        return (_returnType == null) ? o : _returnType.cast(o);
+    }
+
+    @Override
+    public MethodHolder bindTo(Object arg, int idx)
+    {
+        int internalIndex = getInternalIndex(idx);
+        _parameters[internalIndex] = arg;
+        _boundParams[internalIndex] = true;
+        return this;
+    }
+
+    @Override
+    public MethodHolder bindTo(Object arg)
+    {
+        return bindTo(arg, 0);
+    }
+
+    @SuppressWarnings({"UnusedReturnValue", "SameParameterValue"})
+    private MethodHolder insertArguments(boolean bind, int idx, Object... args)
+    {
+        int index = 0;
+        int argsIndex = 0;
+        for (int i = 0; i < _numParams; i++)
+        {
+            if (_boundParams[i] != Boolean.TRUE)
+            {
+                if (index >= idx && argsIndex < args.length)
+                {
+                    Object val = args[argsIndex++];
+                    _parameters[i] = val;
+                    _boundParams[i] = bind;
+                }
+                else
+                {
+                    _parameters[i] = null;
+                }
+                index++;
+            }
+        }
+
+        if (argsIndex < args.length)
+            throw new WrongMethodTypeException(String.format("Expected %s params but had %s", args.length, argsIndex + 1));
+        if (index != argsIndex)
+            throw new WrongMethodTypeException(String.format("Expected %s params but had %s", index + 1, argsIndex + 1));
+        return this;
+    }
+
+    @Override
+    public MethodHolder filterReturnValue(MethodHandle filter)
+    {
+        _returnFilter = filter;
+        return this;
+    }
+
+    @Override
+    public MethodHolder changeReturnType(Class<Object> objectClass)
+    {
+        _returnType = objectClass;
```

Comment:
This is also not correct.   With `MethodHandler` you can interleave multiple calls to filterReturnValue and changeReturnType and they will all apply.    You need to implement this as a filter, so it can be done in order with the other filters.